### PR TITLE
Dev Deploy

### PR DESCRIPTION
**Pull request by johnc1231**

### CODE REVIEW

**Comment by johnc1231** [comment]
Assigned Dan but feel free to look as well @cseed 

Admittedly I'm not 100% sure that the CLI code is correct because I just changed it to make a request instead of just calling the function and I cannot test it until the endpoint exists.

File: ci/ci/constants.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 GITHUB_CLONE_URL = 'https://github.com/'
 
-with open(os.environ['HAIL_TOKEN_FILE']) as f:
+with open(os.environ.get('HAIL_TOKEN_FILE') or os.path.expanduser('~/.hail/token')) as f:
```
Comment (danking) [inline ci/ci/constants.py]: this now appears in aioclient and here. Let's pull it into gear.

```python
def find_token(token_file=None):
    token_file = (token_file or
                  os.environ.get('HAIL_TOKEN_FILE') or
                  os.path.expanduser('~/.hail/token'))
    if not os.path.exists(token_file):
       raise ValueError(
           f'Cannot create a client without a token. No file was '
           f'found at {token_file}')
    with open(token_file) as f:
        return f.read()

def find_userdata(token_file=None):
    return hj.JWTClient.unsafe_decode(find_token(token_file))
```

maybe add them to JWTClient.

File: ci/ci/github.py [hunk 2]
```diff
@@ -768,3 +768,84 @@ def checkout_script(self):
...
+(cd {shq(repo_dir)}; {self.checkout_script()})
+''')
+            with open(f'{repo_dir}/build.yaml', 'r') as f:
+                config = BuildConfiguration(self, f.read(), scope='dev', profile=profile_steps)
+
+            log.info(f'creating dev deploy batch for {self.branch.short_str()} and user {self.user}')
+
+            #Think: For every step, need to know if it had a yaml config to tear down.
+            #So much harder than dropping the namespace. Sigh. 
+            #
```
Comment (danking) [inline ci/ci/github.py]: what's this about?

File: ci/ci/github.py [hunk 3]
```diff
@@ -768,3 +768,84 @@ def checkout_script(self):
...
+                    'user': self.user
+                })
+            config.build(deploy_batch, self, scope='dev')
+            deploy_batch = await deploy_batch.submit()
+            self.deploy_batch = deploy_batch
+            await deploy_batch.wait()
+        except concurrent.futures.CancelledError:
+            raise
+        except Exception as e:  # pylint: disable=broad-except
+            raise e
```
Comment (danking) [inline ci/ci/github.py]: these last two don't seem to matter? you just re-raise the caught exception

File: hail/python/hailtop/hailctl/dev/cli.py [hunk 4]
```diff
@@ -14,20 +16,34 @@ def print_help():
...
     else:
         module = args[0]
-        args = args[1:]
+        extra_args = args[1:]
         if module == 'benchmark':
             from .benchmark import cli
+            cli.main(extra_args)
+        elif module == 'deploy':
+            from .deploy import cli
+            print(args)
```
Comment (danking) [inline hail/python/hailtop/hailctl/dev/cli.py]: probably don't want this

File: hail/python/hailtop/hailctl/dev/deploy/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from . import cli
```
Comment (danking) [inline hail/python/hailtop/hailctl/dev/deploy/__init__.py]: does this have any effect?

File: hail/python/hailtop/hailctl/dev/deploy/cli.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+def init_parser(parser):
+    parser.add_argument('repo', type=str)
+    parser.add_argument('branch', type=str)
+    parser.add_argument('namespace', type=str)
+    parser.add_argument('profile', type=str, choices=['batch_test'])
+
+def main(args):
+    # For now, let's find CI and just manually call the damn function.
+    async with aiohttp.ClientSession() as session:
+        session.post('https://batch.hail.is/api/v1alpha/dev_test_branch/'', json={
```
Comment (danking) [inline hail/python/hailtop/hailctl/dev/deploy/cli.py]: extra single quote?

File: hail/python/hailtop/hailctl/dev/deploy/cli.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import ci
+
+def init_parser(parser):
+    parser.add_argument('repo', type=str)
+    parser.add_argument('branch', type=str)
+    parser.add_argument('namespace', type=str)
+    parser.add_argument('profile', type=str, choices=['batch_test'])
+
+def main(args):
+    # For now, let's find CI and just manually call the damn function.
```
Comment (danking) [inline hail/python/hailtop/hailctl/dev/deploy/cli.py]: no comments

File: hail/python/hailtop/hailctl/dev/deploy/cli.py (hunk 6 above)
Comment (danking) [inline hail/python/hailtop/hailctl/dev/deploy/cli.py]: shouldn't this be ci.hail.is?

File: ci/ci/build.py [hunk 8]
```diff
@@ -524,6 +523,17 @@ def build(self, batch, code, deploy):  # pylint: disable=unused-argument
   apiGroup: ""
 '''
 
+        if scope in ['deploy', 'test']:
+            # FIXME label
+            config = config + f'''\
```
[the remark this answers is not in the document]
Comment (johnc1231) [reply ci/ci/build.py]: I don't think it does but admittedly it's unclear to read so I rearranged.

**Pull request merged**